### PR TITLE
Feature/convert sample app to template

### PR DESCRIPTION
**Pull request by mwoodffdc**
Convert the sample app to a template that can be deployed with the tool.

If there are issues, I'll deal with any in a subsequent (smaller) PR.

### CODE REVIEW

File: templates/sample-app/.config.js [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+
+function conf(name) {
+	return config.get(name) || config.get(name.split('.').join('-'))
+}
+
+var name = package.name || basename
+var spec = npa(name)
+var scope = conf('scope')
+if (scope) {
+	if (scope.charAt(0) !== '@') scope = '@' + scope
```
Comment (dfruddffdc) [inline templates/sample-app/.config.js]: Hmm.... single-line if blocks. :/

File: templates/sample-app/.config.js [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+	return er
+})
+
+var version = package.version || conf('init.version') || '1.0.0'
+exports.version = yes ?
+	version :
+	prompt('version', version, function (version) {
+		if (semver.valid(version)) return version
+		var er = new Error('Invalid version: "' + version + '"')
+		er.notValid = true
```
Comment (dfruddffdc) [inline templates/sample-app/.config.js]: Probably `er.valid = false` will lead to fewer double-negatives down the line.

File: templates/sample-app/.config.js [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+		return er
+	})
+
+if (!package.description) {
+	exports.description = yes ? '' : prompt('description')
+}
+
+if (!package.bin) {
+	exports.bin = function (cb) {
+		fs.readdir(path.resolve(dirname, 'bin'), function (er, d) {
```
Comment (dfruddffdc) [inline templates/sample-app/.config.js]: Abbreviations are evil. I can just about stand `cb`. What's an `er`?
I can guess, but I just don't wanna.

File: templates/sample-app/.config.js [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+			// just take the first js file we find there, or nada
+			return cb(null, d.filter(function (f) {
+				return f.match(/\.js$/)
+			})[0])
+		})
+	}
+}
+
+exports.directories = function (cb) {
+	fs.readdir(dirname, function (er, dirs) {
```
Comment (dfruddffdc) [inline templates/sample-app/.config.js]: Argh, I don't think there's a line in this function that doesn't have an abbreviation in it.
I can take `cb`, `dirs` and `dirname` as they're fairly standard, but would prefer real names.
I don't like `er`, `d` or `res`.

File: templates/sample-app/.config.js [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+		"text",
+		"text-summary"
+	],
+	"cache": true,
+	"all": true
+}
+
+if (!package.repository) {
+	exports.repository = function (cb) {
+		fs.readFile('.git/config', 'utf8', function (er, gconf) {
```
Comment (dfruddffdc) [inline templates/sample-app/.config.js]: Not sure it needs to be quite so clever... each neat time-saver is admittedly neat, but I worry it adds a potential point of failure, e.g. if the user isn't using git for source control.
I'd say the user should just enter their repo details manually, and/or we put a massive try/catch around the whole nice-to-have functionality.
(Something for the backlog).

File: templates/sample-app/res/.settings/org.eclipse.core.resources.prefs [hunk 6]
```diff
@@ -0,0 +1,7 @@
+eclipse.preferences.version=1
```
Comment (dfruddffdc) [inline templates/sample-app/res/.settings/org.eclipse.core.resources.prefs]: Does this therefore imply I have to use Eclipse? What if I want to stick to good ol' vscode?
(Something for the backlog)

File: templates/sample-app/res/src/apex/app/main/default/aura/DataGeneratorButton/DataGeneratorButton.css [hunk 7]
```diff
@@ -0,0 +1,3 @@
+.THIS header {
+    padding-top: 0;
+}
```
Comment (dfruddffdc) [inline templates/sample-app/res/src/apex/app/main/default/aura/DataGeneratorButton/DataGeneratorButton.css]: Whoops, no newline.

File: templates/sample-app/res/src/apex/app/main/default/aura/DataGeneratorStepMonitor/DataGeneratorStepMonitorController.js [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+		$A.enqueueAction(action);
+
+	},
+
+	handleStepEvent: function (component, event, helper) {
+
+		var status = event.getParam('status'),
+			eventId = event.getParam('id'),
+			objectId = component.get('v.recordId'),
```
Comment (dfruddffdc) [inline templates/sample-app/res/src/apex/app/main/default/aura/DataGeneratorStepMonitor/DataGeneratorStepMonitorController.js]: Why did objectId become recordId?

File: templates/sample-app/res/src/apex/app/main/default/classes/DataGeneratorStepTriggerTest.cls [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+ *  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ *  OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ *  OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ **/
+
+@isTest(SeeAllData=false)
+private class DataGeneratorStepTriggerTest 
+{
+    @isTest static void dataGeneratorStepTrigger_shouldUpdateGenerateDataTask()
```
Comment (dfruddffdc) [inline templates/sample-app/res/src/apex/app/main/default/classes/DataGeneratorStepTriggerTest.cls]: Indentation troubles

File: templates/sample-app/res/src/apex/app/main/default/classes/DeliveryRouteMapChooserController.cls [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+
+public with sharing class DeliveryRouteMapChooserController 
+{
+    @auraEnabled
+    public static List<DeliveryRoute__c> getRoutes(List<Id> planIds)
+    {
+		return DeliveryRouteSelector.newInstance().selectByPlanIds(new Set<Id>(planIds));
+    }
+	
+	@AuraEnabled
```
Comment (dfruddffdc) [inline templates/sample-app/res/src/apex/app/main/default/classes/DeliveryRouteMapChooserController.cls]: Indentation and capitalisation troubles

File: templates/sample-app/res/src/apex/app/main/default/classes/DeliveryRouteMapChooserControllerTest.cls [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+ *  OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ *  OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ **/
+
+@isTest(SeeAllData=false)
+private class DeliveryRouteMapChooserControllerTest
+{
+    @isTest
+    static void getRoutes_withPlanId_callsSelector()
```
Comment (dfruddffdc) [inline templates/sample-app/res/src/apex/app/main/default/classes/DeliveryRouteMapChooserControllerTest.cls]: No access level keyword troubles.

File: templates/sample-app/res/src/apex/app/main/default/classes/DeliveryWaypointController.cls [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+ *  THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ *  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ *  OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ *  OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ **/
+
+public with sharing class DeliveryWaypointController 
+{
+    @auraEnabled
```
Comment (dfruddffdc) [inline templates/sample-app/res/src/apex/app/main/default/classes/DeliveryWaypointController.cls]: Inconsistent tabs v spaces troubles.

File: templates/sample-app/res/src/apex/app/main/default/classes/RouteAPITest.cls [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+			System.assert(false, 'Expected exception');
+		}
+		catch(Exception ex)
+		{
+		}
+    }
+
+	@isTest
+	private static void saveRoutes_validRoutes_Saves()
+    {
```
Comment (dfruddffdc) [inline templates/sample-app/res/src/apex/app/main/default/classes/RouteAPITest.cls]: The war between tabs and spaces rages on.

File: templates/sample-app/res/src/apex/app/main/default/classes/RouteAPITest.cls [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+		insert plan;
+		
+		List<DeliveryRoute__c> routes = new List<DeliveryRoute__c>();
+		routes.add(new DeliveryRoute__c(DeliveryPlan__c = plan.Id, URL__c ='Test', Vehicle__c = vehicle.Id));
+		routes.add(new DeliveryRoute__c(DeliveryPlan__c = plan.Id, URL__c ='Test', Vehicle__c = vehicle.Id));
+		routes.add(new DeliveryRoute__c(DeliveryPlan__c = plan.Id, URL__c ='Test', Vehicle__c = vehicle.Id));
+
+		return routes;
+
+	/*	List<DeliveryWaypoint__c> waypoints = new List<DeliveryWaypoint__c>();
```
Comment (dfruddffdc) [inline templates/sample-app/res/src/apex/app/main/default/classes/RouteAPITest.cls]: We don't need no commented out code.

File: templates/sample-app/res/src/apex/app/main/default/objects/Account/listViews/All.listView-meta.xml [hunk 15]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (dfruddffdc) [inline templates/sample-app/res/src/apex/app/main/default/objects/Account/listViews/All.listView-meta.xml]: This is just the default list view. This shouldn't be in source control.

File: templates/sample-app/res/src/apex/app/main/default/objects/Contact/fields/AccountId.field-meta.xml [hunk 16]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (dfruddffdc) [inline templates/sample-app/res/src/apex/app/main/default/objects/Contact/fields/AccountId.field-meta.xml]: All these default Contact fields don't need to be in source control.

File: templates/sample-app/res/src/java/main/java/com/financialforce/routesolver/QuestionConsumer.java [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+
+	public QuestionConsumer(Channel channel) {
+		super(channel, Question.class.getName(), Answer.class.getName());
+		this.solver = new Solver();
+	}
+
+	@Override
+	public Answer handleMessage(Context context, Question question) throws HandleMessageException {
+
+		if (solver == null) {
```
Comment (dfruddffdc) [inline templates/sample-app/res/src/java/main/java/com/financialforce/routesolver/QuestionConsumer.java]: The only constructor sets the solver. Mark it as final, then you won't have to validate it's not null here.

File: templates/web-template/res/src/node/spec/boilerplate/schema.spec.js [hunk 18]
```diff
@@ -0,0 +1,151 @@
...
+
+	describe('getWebSchemas', () => {
+
+		it('should return all the web schemas', () => {
+
+			// Given
+			const expectedOutput = {
+				calculateRoutesForPlan: {
+					fileName: 'calculateRoutesForPlan',
+					path: '/Users/Guest/GIT/test/src/node/lib/schema/api/calculateRoutesForPlan.avsc',
```
Comment (dfruddffdc) [inline templates/web-template/res/src/node/spec/boilerplate/schema.spec.js]: Where did this path come from? Will it work on other machines?

File: templates/sample-app/res/src/node/lib/handler/createData/sampleData/Contact.json [hunk 19]
```diff
@@ -0,0 +1,4119 @@
+{
+	"records": [{
+		"FirstName": "Mission High School",
```
Comment (dfruddffdc) [inline templates/sample-app/res/src/node/lib/handler/createData/sampleData/Contact.json]: A Contact called Mission High School Mission High School? Mean parents! :P

By the way, First Name is optional, Last Name is required, you can just leave the First Names blank and remove some duplication.

File: templates/sample-app/.config.js (hunk 1 above)
Comment (mwoodffdc) [reply templates/sample-app/.config.js]: I'm going to complete drop the `.config.js` file in preference of a `.config.json` file which only contains the bits of this file that we need.

I'll do this in the next P.R. though.

File: templates/sample-app/res/.settings/org.eclipse.core.resources.prefs (hunk 6 above)
Comment (mwoodffdc) [reply templates/sample-app/res/.settings/org.eclipse.core.resources.prefs]: No, it doesn't.

VSCode uses Eclipse plugins thus the settings files are the same.

**Pull request merged**